File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.cpp
class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        // Making the graph in the form of an adjacency matrix
        vector<vector<int>> graph(n, vector<int>(n, INT_MAX));
        
        for(int i = 0; i < n; i++){
            graph[i][i] = 0;
        }

        for(const auto &edge : edges){
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];

            graph[u][v] = w;
            graph[v][u] = w;
        }

        for(int via = 0; via < n; via++){
            for(int i = 0; i < n; i++){
                for(int j = 0; j < n; j++){
                    if(graph[i][via] != INT_MAX and graph[via][j] != INT_MAX){
                        graph[i][j] = min(graph[i][j], graph[i][via] + graph[via][j]);
                    }
                } 
            }
        }

        int minNeighboursCount = INT_MAX;
        int cityWithSmallestNumberOfNeighbors = n;
        
        
        for(int i = 0; i < n; i++){
            int neighbourCount = 0;
            for(int j = 0; j < n; j++){
                if(graph[i][j] <= distanceThreshold) neighbourCount++;
            } 

            if(neighbourCount <= minNeighboursCount){
                minNeighboursCount = neighbourCount;
                cityWithSmallestNumberOfNeighbors = i;
            }
        }
        

        return cityWithSmallestNumberOfNeighbors;

    }
}; 

File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
class Solution {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] shortestPaths = new int[n][n];


        for(int i = 0; i < n; i++){
            Arrays.fill(shortestPaths[i], (int)1e9);

            shortestPaths[i][i] = 0;
        }

        for(int[] edge : edges){
            int u = edge[0], v = edge[1], w = edge[2];

            shortestPaths[u][v] = w;
            shortestPaths[v][u] = w;
        }

        for(int k = 0; k < n; k++){
            for(int i = 0; i < n; i++){
                for(int j = 0; j < n; j++){
                    shortestPaths[i][j] = Math.min(
                        shortestPaths[i][j],
                        shortestPaths[i][k] + shortestPaths[k][j]
                    );
                }
            }
        }

        int minimumCityCount = n;
        int resultCity = 0;

        for(int city = 0; city < n; city++){
            int cityCount = 0;
            for(int i = 0; i < n; i++){
                if(city == i) continue;

                if(shortestPaths[city][i] <= distanceThreshold) cityCount++;
            }

            if(cityCount <= minimumCityCount){
                resultCity = city;
                minimumCityCount = cityCount;
            }
        }

        // System.out.println(Arrays.deepToString(shortestPaths));

        return resultCity;
        
    }
}